src/schema-generator/types.ts

```diff
@@ -0,0 +1,81 @@
+import * as Yup from "yup";
+import { INumericSchema, ITextSchema } from "../fields";
+
+export const SCHEMA_TYPES = ["string", "number", "boolean", "array", "object"] as const;
+export const CONDITIONS = [
+	"required",
+	"length",
+	"min",
+	"max",
+	"matches",
+	"email",
+	"url",
+	"uuid",
+	"positive",
+	"negative",
+	"integer",
+	"lessThan",
+	"moreThan",
+	"when",
+] as const;
+export type TYupSchemaType = (typeof SCHEMA_TYPES)[number];
+export type TCondition = (typeof CONDITIONS)[number];
+
+/** base validation rule to extend from */
+interface IRule {
+	length?: number | undefined;
+	min?: number | undefined;
+	max?: number | undefined;
+	matches?: string | undefined;
+	email?: boolean | undefined;
+	url?: boolean | undefined;
+	uuid?: boolean | undefined;
+	positive?: boolean | undefined;
+	negative?: boolean | undefined;
+	integer?: boolean | undefined;
```

Comment:
Should type-specific rules be moved to the individual field schemas? My main concern is being able to specify validation rules that don't make sense together like `fieldType=string` and `positive=true`.

Alternatively we can provide a better error message for the developer to fix the rule. At the moment since the function is undefined, there will be this error: `TypeError: yupSchema[condition] is not a function`

Reply:
yeah that's a valid point, I was intending to keep the definitions simple first and agree we should type it properly. I have added a TODO comment for this and added more descriptive errors `mapRules()`.

in the long run, we will need to make the change on both this library and Frontend Engine, I have created a ticket in the mural board to track this.